File: tests/regressions/uimd/parity/python/stale_scrollview_focus/stale_scrollview_focus.py
"""Manual regression app for replacing a focused ScrollView page."""

import os
import sys
from dataclasses import dataclass


_HERE = os.path.dirname(os.path.abspath(__file__))
_ROOT = os.path.abspath(os.path.join(_HERE, "..", "..", "..", "..", "..", ".."))
sys.path.insert(0, os.path.join(_ROOT, "src"))
sys.path.insert(0, _HERE)

from edit_field_row_ui import EditFieldRowUI
from edit_fields_scroll_ui import EditFieldsScrollUI
from edit_page_ui import EditPageUI
from item_row_ui import ItemRowUI
from list_page_ui import ListPageUI
from stale_scrollview_focus_ui import StaleScrollviewFocusUI
from uimd.runtime import UIApplication


MINIMUM_RENDER_HEIGHT = 1
FIELD_COUNT = 40
PRIMARY_FIELD_INDEX = 0
BODY_FIELD_INDEX = 1
DISPLAY_INDEX_OFFSET = 1
FIELD_NUMBER_WIDTH = 2


@dataclass(frozen=True)
class ReproItem:
    name: str
    summary: str
    body: str


ITEMS = (
    ReproItem(
        "Alpha Apartments",
        "First row, visible without scrolling",
        "Alpha item was opened from the top of the ScrollView.\n"
        "The edit page should not keep any stale list focus background.",
    ),
    ReproItem(
        "Bravo Offices",
        "Second row, used by the automated repro",
        "Bravo item was opened by pressing Enter on a row button.\n"
        "The old ScrollView page must be gone before this page renders.",
    ),
    ReproItem(
        "Charlie Lofts",
        "Third row, keeps the source list taller",
        "Charlie item keeps the source page closer to the reported list shape.\n"
        "The opened edit page still owns the new nested ScrollView.",
    ),
    ReproItem(
        "Delta Studios",
        "Fourth row, manual repro filler",
        "Delta item is extra manual test data for scroll and focus traversal.\n"
        "Activating Open should replace the whole page cleanly.",
    ),
    ReproItem(
        "Echo Retail",
        "Fifth row, enough rows to browse",
        "Echo item makes the source ScrollView visibly list-like.\n"
        "The edit form below contains many alternating controls.",
    ),
)


def field_number(index):
    return str(index + DISPLAY_INDEX_OFFSET).zfill(FIELD_NUMBER_WIDTH)


def field_value(item, index):
    if index == PRIMARY_FIELD_INDEX:
        return f"{item.name} primary field"
    if index == BODY_FIELD_INDEX:
        return item.body.splitlines()[PRIMARY_FIELD_INDEX]
    return f"{item.name} detail field {field_number(index)}"


def render_expand_width_child(child, width):
    _natural_width, natural_height = child.get_natural_size()
    child.resize(width, max(MINIMUM_RENDER_HEIGHT, natural_height))
    return child.render()


class ItemRow(ItemRowUI):
    """One interactive row inside the regression ScrollView."""

    def __init__(self, item, on_open):
        super().__init__()
        self._item = item
        self._on_open = on_open

    def open(self):
        super().open()
        self.name.text = self._item.name
        self.summary.text = self._item.summary

    def on_open_btn_click(self):
        self._on_open(self._item)


class ListPage(ListPageUI):
    """ScrollView page whose row button replaces the ViewHost page."""

    def __init__(self, items, on_open):
        super().__init__()
        self._items = items
        self._on_open = on_open

    def open(self):
        super().open()
        self.populate()

    def populate(self):
        self.clear_children()
        for item in self._items:
            row = ItemRow(item, self._on_open)
            row.parent = self
            row._app = self._app
            row.mode = "expand_width"
            row.open()
            self.add_child(row)
        self.scroll_to_top()

    def _render_child(self, child, width):
        return render_expand_width_child(child, width)


class EditFieldRow(EditFieldRowUI):
    """One text input followed by an action button in the edit ScrollView."""

    def __init__(self, index, value, on_action):
        super().__init__()
        self._index = index
        self._value = value
        self._on_action = on_action

    def open(self):
        super().open()
        number = field_number(self._index)
        self.field_input.value = self._value
        self.field_btn.title = f"Action {number}"

    def on_field_btn_click(self):
        self._on_action(self._index, self.field_input.value)


class EditFieldsScroll(EditFieldsScrollUI):
    """Large edit form body with many alternating text inputs and buttons."""

    def __init__(self, item, on_action):
        super().__init__()
        self._item = item
        self._on_action = on_action

    def open(self):
        super().open()
        self.populate()

    def populate(self):
        self.clear_children()
        for index in range(FIELD_COUNT):
            row = EditFieldRow(index, field_value(self._item, index), self._on_action)
            row.parent = self
            row._app = self._app
            row.mode = "expand_width"
            row.open()
            self.add_child(row)
        self.scroll_to_top()

    def _render_child(self, child, width):
        return render_expand_width_child(child, width)


class EditPage(EditPageUI):
    """Edit page displayed after the ScrollView row button fires."""

    def __init__(self, item, on_back):
        super().__init__()
        self._item = item
        self._on_back = on_back

    def open(self):
        super().open()
        self.edit_title.text = f"Edit {self._item.name}"
        self.name_input.value = self._item.name
        self.fields.set_view(EditFieldsScroll(self._item, self._field_action))

    def on_back_btn_click(self):
        self._on_back()

    def on_save_btn_click(self):
        self.status.text = "Saved in memory"

    def _field_action(self, index, value):
        self.status.text = f"Action {field_number(index)}: {value}"


class StaleScrollviewFocus(StaleScrollviewFocusUI):
    """Root app that swaps a ViewHost from list page to edit page."""

    def open(self):
        super().open()
        self.show_list_page()

    def show_list_page(self):
        self.page.set_view(ListPage(ITEMS, self.show_edit_page))
        self.hint.text = "Focus list, press Enter, choose Open."

    def show_edit_page(self, item):
        self.page.set_view(EditPage(item, self.show_list_page))
        self.hint.text = "Edit page opened from ScrollView row."

    def on_tab_list_click(self):
        self.show_list_page()

    def on_tab_other_click(self):
        self.hint.text = "Other tab intentionally leaves the list unchanged."

    def on_quit_btn_click(self):
        if self._app is not None:
            self._app._running = False


def main():
    app = UIApplication()
    app.open(StaleScrollviewFocus())
    return app.run()


if __name__ == "__main__":
    sys.exit(main())

File: tests/regressions/uimd/parity/cpp/stale_scrollview_focus/stale_scrollview_focus.cpp
#include "edit_page_ui.hpp"
#include "edit_field_row_ui.hpp"
#include "edit_fields_scroll_ui.hpp"
#include "item_row_ui.hpp"
#include "list_page_ui.hpp"
#include "stale_scrollview_focus_ui.hpp"

#include "ui/elements/ReusableElement.hpp"
#include "ui/generated/GeneratedWindowRuntime.hpp"

#include <algorithm>
#include <array>
#include <cstddef>
#include <functional>
#include <memory>
#include <string>
#include <string_view>
#include <utility>

namespace
{

constexpr int kMinimumRenderHeight = 1;
constexpr std::size_t kFieldCount = 40;
constexpr std::size_t kPrimaryFieldIndex = 0;
constexpr std::size_t kBodyFieldIndex = 1;
constexpr std::size_t kDisplayIndexOffset = 1;
constexpr std::size_t kFieldNumberPaddingThreshold = 10;

struct ReproItem
{
    std::string_view name;
    std::string_view summary;
    std::string_view body;
};

constexpr std::array<ReproItem, 5> kItems{{
    {
        "Alpha Apartments",
        "First row, visible without scrolling",
        "Alpha item was opened from the top of the ScrollView.\n"
        "The edit page should not keep any stale list focus background.",
    },
    {
        "Bravo Offices",
        "Second row, used by the automated repro",
        "Bravo item was opened by pressing Enter on a row button.\n"
        "The old ScrollView page must be gone before this page renders.",
    },
    {
        "Charlie Lofts",
        "Third row, keeps the source list taller",
        "Charlie item keeps the source page closer to the reported list shape.\n"
        "The opened edit page still owns the new nested ScrollView.",
    },
    {
        "Delta Studios",
        "Fourth row, manual repro filler",
        "Delta item is extra manual test data for scroll and focus traversal.\n"
        "Activating Open should replace the whole page cleanly.",
    },
    {
        "Echo Retail",
        "Fifth row, enough rows to browse",
        "Echo item makes the source ScrollView visibly list-like.\n"
        "The edit form below contains many alternating controls.",
    },
}};

using OpenCallback = std::function<void(const ReproItem&)>;
using BackCallback = std::function<void()>;
using FieldActionCallback = std::function<void(std::size_t, std::string_view)>;

[[nodiscard]] std::string fieldNumber(std::size_t index)
{
    const std::size_t displayIndex = index + kDisplayIndexOffset;
    std::string value = std::to_string(displayIndex);
    if (displayIndex < kFieldNumberPaddingThreshold)
    {
        value = "0" + value;
    }
    return value;
}

[[nodiscard]] std::string firstLine(std::string_view value)
{
    const std::size_t lineBreak = value.find('\n');
    return std::string{value.substr(0, lineBreak)};
}

[[nodiscard]] std::string fieldValue(const ReproItem& item, std::size_t index)
{
    if (index == kPrimaryFieldIndex)
    {
        return std::string{item.name} + " primary field";
    }
    if (index == kBodyFieldIndex)
    {
        return firstLine(item.body);
    }
    return std::string{item.name} + " detail field " + fieldNumber(index);
}

template <typename RowWindow>
void addReusableScrollChild(ui::GeneratedScrollViewBase& scrollView, std::unique_ptr<RowWindow> row)
{
    const ui::Size naturalSize = ui::generatedWindowContentSize(*row);
    auto reusable = std::make_unique<ui::ReusableElement>("row", std::move(row));
    reusable->setFrame(ui::Rect{0, 0, 0, std::max(kMinimumRenderHeight, naturalSize.height)});
    scrollView.addChild(std::move(reusable));
}

class ItemRowWindow : public ItemRowUI
{
public:
    ItemRowWindow(ReproItem item, OpenCallback onOpen)
        : item_(item),
          onOpen_(std::move(onOpen))
    {
        name->setText(std::string{item_.name});
        summary->setText(std::string{item_.summary});
    }

    void onOpenBtnClick() override
    {
        if (onOpen_)
        {
            onOpen_(item_);
        }
    }

private:
    ReproItem item_;
    OpenCallback onOpen_;
};

class ListPageWindow : public ListPageUI
{
public:
    explicit ListPageWindow(OpenCallback onOpen)
        : onOpen_(std::move(onOpen))
    {
        populate();
    }

private:
    void populate()
    {
        clearChildren();
        for (const ReproItem& item : kItems)
        {
            addReusableScrollChild(*this, std::make_unique<ItemRowWindow>(item, onOpen_));
        }
        const ui::Rect currentFrame = frame();
        if (currentFrame.width > 0 && currentFrame.height > 0)
        {
            (void)scrollToTop(ui::Size{currentFrame.width, currentFrame.height});
        }
    }

    OpenCallback onOpen_;
};

class EditFieldRowWindow : public EditFieldRowUI
{
public:
    EditFieldRowWindow(std::size_t index, std::string value, FieldActionCallback onAction)
        : index_(index),
          onAction_(std::move(onAction))
    {
        field_input->setValue(std::move(value));
        field_btn->setTitle("Action " + fieldNumber(index_));
    }

    void onFieldBtnClick() override
    {
        if (onAction_)
        {
            onAction_(index_, field_input->value());
        }
    }

private:
    std::size_t index_;
    FieldActionCallback onAction_;
};

class EditFieldsScrollWindow : public EditFieldsScrollUI
{
public:
    EditFieldsScrollWindow(ReproItem item, FieldActionCallback onAction)
        : item_(item),
          onAction_(std::move(onAction))
    {
        populate();
    }

private:
    void populate()
    {
        clearChildren();
        for (std::size_t index = 0; index < kFieldCount; ++index)
        {
            addReusableScrollChild(
                *this,
                std::make_unique<EditFieldRowWindow>(index, fieldValue(item_, index), onAction_)
            );
        }
        const ui::Rect currentFrame = frame();
        if (currentFrame.width > 0 && currentFrame.height > 0)
        {
            (void)scrollToTop(ui::Size{currentFrame.width, currentFrame.height});
        }
    }

    ReproItem item_;
    FieldActionCallback onAction_;
};

class EditPageWindow : public EditPageUI
{
public:
    EditPageWindow(ReproItem item, BackCallback onBack)
        : onBack_(std::move(onBack))
    {
        edit_title->setText("Edit " + std::string{item.name});
        name_input->setValue(std::string{item.name});
        fields->setView(std::make_unique<EditFieldsScrollWindow>(
            item,
            [this](std::size_t index, std::string_view value)
            {
                status->setText("Action " + fieldNumber(index) + ": " + std::string{value});
            }
        ));
        auto* fieldScroll = dynamic_cast<EditFieldsScrollWindow*>(fields->child());
        if (fieldScroll != nullptr)
        {
            const ui::Size fieldsSize = ui::generatedWindowContentSize(*fieldScroll);
            if (fieldsSize.width > 0 && fieldsSize.height > 0)
            {
                ui::ScrollViewPosition topPosition;
                topPosition.scrollOffset = 0;
                topPosition.viewOffset = fieldScroll->scrollView().contentHeight(fieldsSize.width);
                topPosition.autoScroll = false;
                fieldScroll->scrollView().restoreScrollPosition(topPosition);
            }
        }
    }

    void onBackBtnClick() override
    {
        if (onBack_)
        {
            onBack_();
        }
    }

    void onSaveBtnClick() override
    {
        status->setText("Saved in memory");
    }

private:
    BackCallback onBack_;
};

class StaleScrollviewFocusApp : public StaleScrollviewFocusUI
{
public:
    StaleScrollviewFocusApp()
    {
        showListPage();
    }

    void onTabListClick() override
    {
        showListPage();
    }

    void onTabOtherClick() override
    {
        hint->setText("Other tab intentionally leaves the list unchanged.");
    }

    void onQuitBtnClick() override
    {
        quitRequested_ = true;
    }

    bool shouldClose() const override
    {
        return quitRequested_;
    }

private:
    void showListPage()
    {
        page->setView(std::make_unique<ListPageWindow>(
            [this](const ReproItem& item)
            {
                showEditPage(item);
            }
        ));
        hint->setText("Focus list, press Enter, choose Open.");
    }

    void showEditPage(const ReproItem& item)
    {
        page->setView(std::make_unique<EditPageWindow>(
            item,
            [this]()
            {
                showListPage();
            }
        ));
        hint->setText("Edit page opened from ScrollView row.");
    }

    bool quitRequested_ = false;
};

}  // namespace

int main(int argc, char** argv)
{
    StaleScrollviewFocusApp app;
    return ui::runGeneratedWindow(app, app.runtimeOptions(), argc, argv);
}
